File: Swift-C++-Testsuite/test/protocols/Swift/PropertyRequirements.swift
@main
struct MyMain {

    static func main() {
        let john = Person(fullName: "John Appleseed")
        var ncc1701 = Starship(name: "Enterprise", prefix: "USS")
    }

}

struct Person: FullyNamed {
    var fullName: String
}

class Starship: FullyNamed {
    var prefix: String?
    var name: String
    init(name: String, prefix: String? = nil) {
        self.name = name
        self.prefix = prefix
    }
    var fullName: String {
        return (prefix != nil ? prefix! + " " : "") + name
    }
}

protocol FullyNamed {
    var fullName: String { get }
}

File: Swift-C++-Testsuite/test/protocols/C++/PropertyRequirements.cpp
#include <iostream>
#include <string>

class FullyNamed {
  public:
    virtual std::string getFullName() = 0;
};

class Person : public FullyNamed {
    public: 
    Person(std::string name) {
      fullName = name;
    }
    std::string getFullName() {
      return fullName;
    }
    std::string fullName;
};

class Starship : public FullyNamed {
  private:
    std::string name;
    std::string prefix;
  public:
    Starship(std::string name, std::string prefix = "") : name(name), prefix(prefix) {}
    std::string getFullName() {
        return (prefix.empty() ? "" : prefix + " ") + name;
    }
};


int main() {
    Person john = {"John Appleseed"};
    Starship ncc1701 = {"Enterprise", "USS"};
    return 0;
}
